File: kth_smallest_sortedMatrix.cpp
// Time Complexity : O(m*n*log(k)) for traversing throught the complete matrix and minheap
// Space Complexity : O(k) for the minheap created
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : no issues as of now.. Learning



class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
       priority_queue <int, vector<int>, greater<int> > myheap;
        int heap_size = (matrix[0].size()*matrix.size())-k+1;
        for(int i=0;i<matrix.size();++i)
        {
            for(int j=0;j<matrix[0].size();++j)
            {
                myheap.push(matrix[i][j]);
                if(myheap.size()>heap_size) myheap.pop();
            }
            
        }
        return myheap.top();
    }
    
};

File: kth-smallest-sorted-matrix.py
# Time Complexity : O(X + Klog(X))
# Space Complexity : O(X)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

import heapq
from typing import List


class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        h = []

        # going through elements int he matrix and appending them to the heap
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                h.append(matrix[i][j])

        heapq.heapify(h)

        # iterating through k range of elements
        for i in range(k - 1):
            heapq.heappop(h)  # popping top elements

        return h[0]  # returning the top element
